llvm/utils/TableGen/X86FoldTablesEmitter.cpp

```diff
@@ -433,7 +433,9 @@ void X86FoldTablesEmitter::addEntryWithFlags(FoldTable &Table,
   // the unfolded load size will be based on the register size. If that’s bigger
   // than the memory operand size, the unfolded load will load more memory and
   // potentially cause a memory fault.
-  if (getRegOperandSize(RegOpRec) > getMemOperandSize(MemOpRec))
+  // And X86 would not unfold RMW instrs.
+  if (getRegOperandSize(RegOpRec) > getMemOperandSize(MemOpRec) ||
+      &Table == &Table2Addr)
```

Comment:
Is that the best we can do? Just bailing based on which table we're emitting seems very coarse.

Reply:
We could add it in outer caller,
`  // Instructions which Read-Modify-Write should be added to Table2Addr.
  if (!MemOutSize && RegOutSize == 1 && MemInSize == RegInSize) {
    addEntryWithFlags(Table2Addr, RegInstr, MemInstr, S | TB_NO_REVERSE, 0, IsManual);
    return;
  }`
Maybe this could be better?

Reply:
That sounds better.

Reply:
I found the S attribute is workable only for manual table, so this doesn't work. So I may need to change the  logic for addEntryWithFlags.